### PR TITLE
feat: Add Poetry

### LINKED ISSUE #609: Specifying project dependencies
Currently, there is a list of general dependencies in the readme for various aspects of trackma, but those don't map cleanly to Python package names on pypi. It would seem rather useful to me to have these in some form, either in `setup.py` as extras or the readme.

At the same time, what do you know of `pyproject.toml` and using e.g. poetry for managing development environments instead of `setup.py`? I've had a rather good experience with those and they definitely appear to be the way to go forward with, ~~although poetry's dependency groups feature with its recent additions is currently not released in any version~~.

### PR DESCRIPTION
**Pull request by txtsd**
This adds dependencies for a virtual environment. 
Removes `pyqt4`.
Updates minimum python version.
Removes PyQt4 since Qt4 has been EOL for several years.

~~Python is currently pinned to 3.7.14 because `pycairo` is not compatible with python 3.6.
Let me know if the versions should be `>=` instead of `==`.~~

Should close #609 when done

EDIT: ~~Hold on I might be missing gtk deps.~~ Nevermind.
EDIT: Additionally, drops `setup.py` in favor of `poetry`, and updates the `README` with relevant info.
EDIT: Python is now `3.7+`

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+[tool.poetry]
+name = "trackma"
+version = "2022.10"
+description = "Open multi-site list manager"
+authors = ["z411 <z411@krutt.org>"]
+license = "GPL-3.0-or-later"
+readme = "README.md"
+
+[tool.poetry.dependencies]
+python = "3.7.14"
```
Comment (FichteFoll) [inline pyproject.toml:10]: This should definitely be `^3.7`. pycairo should be an optional dependency, but [3.6 has reached EOL by almost a year now](https://peps.python.org/pep-0494/) so requiring 3.7 should be fine. That also allows us to use `from __future__ import annotations` and makes way for a lot of useful annotation features that have been added in later versions.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+pygobject = "3.42.1"
+pyinotify = "0.9.6"
+pypresence = "4.2.1"
+pyqt5 = "5.15.7"
+requests = "2.28.1"
+requests-oauthlib = "1.3.1"
+twitter = "1.19.3"
+urwid = "2.1.2"
+setuptools = "63.1.0"
+sip = "6.6.2"
```
Comment (FichteFoll) [inline pyproject.toml:24]: The UI-specific dependencies should be optional and specified with extras.

See https://python-poetry.org/docs/pyproject/#extras

File: setup.py [hunk 3]
```diff
@@ -42,16 +42,18 @@
         'gui_scripts': [
             'trackma-gtk = trackma.ui.gtk:main [GTK]',
             'trackma-qt = trackma.ui.qt:main [Qt]',
-            'trackma-qt4 = trackma.ui.qt.qt4:main [Qt4]',
         ]
     },
```
Comment (FichteFoll) [inline setup.py:46]: Please add these scripts to the poetry file as well.

https://python-poetry.org/docs/pyproject/#scripts

File: README.md [hunk 4]
```diff
@@ -49,7 +49,7 @@ Dependencies
 
 The only required dependency to run Trackma is:
 
-- Python 3.4/3.5
+- Python 3.7.14
```
Comment (FichteFoll) [inline README.md:52]:
```suggestion
- Python 3.7+
```

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+[tool.poetry]
+name = "trackma"
+version = "2022.10"
```
Comment (FichteFoll) [inline pyproject.toml:3]: trackma uses a different versioning scheme.

**Review by FichteFoll — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (txtsd) [reply pyproject.toml]: Done

File: setup.py (hunk 3 above)
Comment (txtsd) [reply setup.py]: Done

File: README.md (hunk 4 above)
Comment (txtsd) [reply README.md]: Done

**Review by FichteFoll — APPROVED** [review]
I haven't tested thislocally but the changes look good to me.

**Comment by txtsd** [comment]
Conflicts resolved.

**Review by HS-157 — APPROVED** [review]
Hi, I am not a expert with Poetry but I tested locally following the instruction from the Readme.
And it's run with differently frontend on Archlinux !
```
❯ python --version
Python 3.10.8
❯ poetry --version
Poetry (version 1.3.1)
```
For me, all it's good, thank you.

```
❯ poetry run trackma-gtk
Trackma GTK v0.8.5
Engine: Trackma v0.8.5 - using account HS157(anilist).
Engine: Reading config files...
Data: Initializing...
libanilist: Initializing...
Data: Using libanilist (anime)
libanilist: Downloading list...
Engine: Parsing redirection file...
Engine: Scanning local library...
Tracker (pyinotify): Initializing...
Tracker (pyinotify): Using pyinotify.
Ready.
Engine: Unloading...
Tracker (pyinotify): Unloading...
Engine: Unloading user hooks...
```
2022-12-16-012502_grim

**Comment by txtsd** [comment]
Wow PEP 582 is YUM!

**Comment by z411** [comment]
Does it support to load the program version from `utils.VERSION`? I like to keep the version in one place as to avoid inconsistencies.

**Comment by txtsd** [comment]
> Does it support to load the program version from `utils.VERSION`? I like to keep the version in one place as to avoid inconsistencies.

I don't understand. Can you elaborate on what you mean?

**Comment by txtsd** [comment]
@z411 There is [poetry-dynamic-versioning](https://pypi.org/project/poetry-dynamic-versioning/) which can set the version, but it only grabs the version from tags. So unless you want to make tags the true source of the version.

If we go that way, it will also require running `poetry dynamic-versioning` (note the space) to set the value in `pyproject.toml` and `utils.py` so they can then be committed. This is because by default, `poetry-dynamic-versioning` sets the version while it's working/building (ideally to a placeholder (`version = "0.0.0"`), then reverts it when it's done.

Another solution is to use a pre-commit hook to block when the version values in `pyproject.toml` and `utils.py` are different.

**Comment by FichteFoll** [comment]
I thought the dynamic versioning plugin and the pre-commit hook were two different approaches to this problem? Imo this can also be done at a later point. I'd merge the PR as soon as I find some time in my day to test the various poetry commands on it locally, but it does look good already.

**Comment by txtsd** [comment]
> I thought the dynamic versioning plugin and the pre-commit hook were two different approaches to this problem? Imo this can also be done at a later point.

Yes, just a little holiday haze 😅

> I'll merge the PR as soon as I find some time in my day to test the various poetry commands on it locally, but it does look good already.

Okay good! I'll look into getting Qt6 support for #628 after that.

**Comment by FichteFoll** [comment]
Finally managed to test this properly (after I fixed the problems I had with poetry&pip always trying to use some keyring that I never configured, https://github.com/pypa/pip/issues/6773, https://github.com/jaraco/keyring#disabling-keyring).

Also ran into an issue with Arch packaging regarding poetry and had to manually downgrade a package there (https://bugs.archlinux.org/task/77375).

I have made some slight adjustments to the readme and merged it into `master`. Note that the readme now does not correctly reflect the extras included on the latest version on pypi until the next version update.

**Comment by txtsd** [comment]
Awesome!

**Pull request merged**